Fix #211 and others

linehill opened the pull request:
* Fixed RegisterdVarCopy() didn’t synchronize pinned host allocations if the pointers were passed indirectly to kernels. Fixed by synchronizing all pinned host allocations known by the runtime.
* Fixed OpenCL and Level Zero backend drivers were not informed about indirect allocations.
* Fixed a test with false positive test outcome.

Fixes #211.
Closes #386.

pjaaskel reviewed (approved):
This looks good to me, but requires thorough testing on dGPUs too. The basic kernel analysis you mention for detecting indirect access could be worthwhile to implement since it should be straightforward and I have a gut feeling it'd handle 99% of current kernels.

linehill commented:
OpenCL backend needs fixing (multi-threading issues in the OpenCL BE).

linehill commented:
* OpenCL: Fixed invalid SVM pointers were annotated in multi-threaded applications.
* OpenCL: Fixed segfaults probably caused by cl_kernel objects being shared among threads.

linehill commented:
Nearly all tests on POCL-CPU fails with `CHIP critical [TID 694387] [1681808645.078377965] : OpenCL failed to initialize any devices` error. Is the installation broken in the CI?

pvelesko commented:
> Nearly all tests on POCL-CPU fails with `CHIP critical [TID 694387] [1681808645.078377965] : OpenCL failed to initialize any devices` error. Is the installation broken in the CI?

I was doing some maintenance earlier this morning and might have not shut off the runner service - I'll just rerun the test before investigating.

pjaaskel commented on src/backend/OpenCL/CHIPBackendOpenCL.cc:
```diff
@@ -938,22 +997,36 @@ CHIPEvent *CHIPQueueOpenCL::launchImpl(CHIPExecItem *ExecItem) {
+    //   shared by exec item, which might get destroyed before the
+    //   kernel is launched/completed
+    //
+    // * Annotated SVM pointers may need to outlive the kernel
```
Isn't it down to the program logic to not free SVM allocations while they might be in use?

linehill replied:
All SVM allocations from all threads are annotated as we can’t tell which ones are needed for the kernel launch. One thread may free a SVM allocation while another thread still holds it.

pjaaskel replied:
OK, I see -- annoying. Did you manage to crash an OpenCL implementation by freeing an SVM allocation while/before the kernel was running? Sounds like an undefined corner case indeed.

pjaaskel replied:
We might want to update the OpenCL spec for this and make it defined behavior: If the SVM alloc is freed during or before, the implementation should ignore it when synchronizing the data.

pjaaskel reviewed (approved):
Other than that, LGTM.

pvelesko commented:
Cancelling CI, need to merge a fix for driver mgmt. Will restart once done

pvelesko commented:
Rebasing on main (CI upgrades) & running CI
